### DISCUSSION TITLE
issue with [@types/ldapjs]

### DISCUSSION THREAD

**Discussion by ShlomitGilo1** [DefinitelyTyped/DefinitelyTyped#57604] [Issues with a @types package]
Hi,
I've installed ldapjs@1.0.2 with @types/ldapjs@1.0.7 in my ts node project.

ldapjs client methods are generally signatured as:
`Client.<method_name>(<arguments...>, **controls**, callback)`
e.g. `bind(dn, password, controls, callback)`

I'm trying to use the controls argument this way:
`Client.bind(<dn>, <password>, {type: "1.3.6.1.4.1.42.2.27.8.5.1", criticality: false}, <callback>)`
but it throws the exception:
`controls must be [Control]`, although I'm using the argument as documented in https://github.com/ldapjs/node-ldapjs/tree/master/docs

I opened an issue in https://github.com/ldapjs/node-ldapjs/issues/770, but they say that:
```
This project is a JavaScript project. If you use it with TypeScript then anything pertaining to TypeScript is up to you.
types are provided by @types/ldapjs. We cannot confirm that they are accurate since we did not write them.
If you have an issue with the types in @types/ldapjs you can create an issue on the DefinitelyTyped repo.
```
In index.d.ts, this argument is defined as any
`export type Control = any;`

I can't figure out how to use this argument.
I can't find any ts example.

[a comment by an automated account is not shown]

**Comment by pmoleri** [comment]
Hi @ShlomitGilo1. You can see [my answer in the original issue](https://github.com/ldapjs/node-ldapjs/issues/770#issuecomment-988323699).
The root issue is that you are not using `new ldap.Control(..)` to create an instance of Control class.

Aside from that:
 - ldapjs docs could be improved to explain this better
 - typescript definitions could be improved that Control is actually a class, instead of an `any` type.

**Reply by ShlomitGilo1** [reply]
I can't use the class, because Control is defined in the ts.d.ts as type, so I get the error:
`'Control' only refers to a type, but is being used as a value here.`

The ldapjs package and the types package are not aligned. The controls attribute cannot be used in ts

**Reply by pmoleri** [reply]
I see. The best would be if you can create a PR in definitely typed with the missing class.
Otherwise, to work around it, you add the missing types right in your project:
Create a `ldapjs-types.d.ts` file anywhere in your project with:
```
// Add missing `Control` class declaration to "ldapjs" module
declare module "ldapjs" {
    export class Control {
        constructor(options: { type: string; criticality?: boolean; value?: any });
    }
}
```
Then you'll be able to import it and use it.
